### PR TITLE
[SPARK-19385][SQL] During canonicalization, `NOT(...(l, r))` should not expect such cases that l.hashcode > r.hashcode

### PR DESCRIPTION
**Pull request by lw-lin**
## What changes were proposed in this pull request?

During canonicalization, `NOT(...(l, r))` should not expect such cases that `l.hashcode > r.hashcode`.

Take the rule `case NOT(GreaterThan(l, r)) if l.hashcode > r.hashcode` for example, it should never be matched since `GreaterThan(l, r)` itself would be re-written as `GreaterThan(r, l)` given `l.hashcode > r.hashcode` after canonicalization.

This patch consolidates rules like `case NOT(GreaterThan(l, r)) if l.hashcode > r.hashcode` and `case NOT(GreaterThan(l, r))`.

## How was this patch tested?

This patch expanded the `NOT` test case to cover both cases where:
- `l.hashcode > r.hashcode`
- `l.hashcode < r.hashcode`

### CODE REVIEW

**Comment by lw-lin** [comment]
@cloud-fan @gatorsmile @dongjoon-hyun would you take a look, thanks!

**Comment by dongjoon-hyun** [comment]
Hi, @lw-lin .
Thank you for pining me. I'll take a look.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala [hunk 1]
```diff
@@ -75,10 +107,14 @@ class ExpressionSetSuite extends SparkFunSuite {
...
-  setTest(1, Not(aUpper >= 1), aUpper < 1, Not(Literal(1) <= aUpper), Literal(1) > aUpper)
-  setTest(1, Not(aUpper <= 1), aUpper > 1, Not(Literal(1) >= aUpper), Literal(1) < aUpper)
+  setTest(1, Not(maxHash > 1), maxHash <= 1, Not(Literal(1) < maxHash), Literal(1) >= maxHash)
+  setTest(1, Not(minHash > 1), minHash <= 1, Not(Literal(1) < minHash), Literal(1) >= minHash)
+  setTest(1, Not(maxHash < 1), maxHash >= 1, Not(Literal(1) > maxHash), Literal(1) <= maxHash)
+  setTest(1, Not(minHash < 1), minHash >= 1, Not(Literal(1) > minHash), Literal(1) <= minHash)
+  setTest(1, Not(maxHash >= 1), maxHash < 1, Not(Literal(1) <= maxHash), Literal(1) > maxHash)
+  setTest(1, Not(minHash >= 1), minHash < 1, Not(Literal(1) <= minHash), Literal(1) > minHash)
+  setTest(1, Not(maxHash <= 1), maxHash > 1, Not(Literal(1) >= maxHash), Literal(1) < maxHash)
+  setTest(1, Not(minHash <= 1), minHash > 1, Not(Literal(1) >= minHash), Literal(1) < minHash)
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala]: These test cases are covered previously correctly. Actually, this PR simplifies the logics only. Am I right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala [hunk 2]
```diff
@@ -78,14 +78,18 @@ object Canonicalize extends {
...
-    case Not(GreaterThan(l, r)) if l.hashCode() > r.hashCode() => GreaterThan(r, l)
-    case Not(GreaterThan(l, r)) => LessThanOrEqual(l, r)
-    case Not(LessThan(l, r)) if l.hashCode() > r.hashCode() => LessThan(r, l)
-    case Not(LessThan(l, r)) => GreaterThanOrEqual(l, r)
-    case Not(GreaterThanOrEqual(l, r)) if l.hashCode() > r.hashCode() => GreaterThanOrEqual(r, l)
-    case Not(GreaterThanOrEqual(l, r)) => LessThan(l, r)
-    case Not(LessThanOrEqual(l, r)) if l.hashCode() > r.hashCode() => LessThanOrEqual(r, l)
-    case Not(LessThanOrEqual(l, r)) => GreaterThan(l, r)
+    case Not(GreaterThan(l, r)) =>
+      assert(l.hashCode() <= r.hashCode())
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala]: Can we remove these `assert`s? It seems to be verified with your test cases now.

**Comment by dongjoon-hyun** [comment]
The original logic was designed to be safe for changing the caller bottom-up code, [here](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala#L154-L157).

```scala
  lazy val canonicalized: Expression = {
    val canonicalizedChildren = children.map(_.canonicalized)
    Canonicalize.execute(withNewChildren(canonicalizedChildren))
  }
```
But, I agree that it's safe to simplify that with the new @lw-lin 's test cases.

For the `assert` statements, I think @cloud-fan and @gatorsmile can give more insightful advice.

For me, LGTM except that.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala (hunk 1 above)
Comment (lw-lin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala]: yea sure they are covered correctly even prior to this patch's changes!

the previous `aUpper`'hashcode is either greater than or less than `1`'s hashcode but can not be both, while this change aims to test both cases -- but I'm quite open to revert the changes if they are considered unnecessary.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala (hunk 2 above)
Comment (lw-lin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala]: thanks! maybe an alternative way is to add comments saying it's guaranteed that `l.hashcode <= r.hashcode`, otherwise people might wonder why there is no `case Not(LessThanOrEqual(l, r)) if l.hashCode() > r.hashCode()` at their first glance.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala [hunk 3]
```diff
@@ -78,14 +78,18 @@ object Canonicalize extends {
     case GreaterThanOrEqual(l, r) if l.hashCode() > r.hashCode() => LessThanOrEqual(r, l)
     case LessThanOrEqual(l, r) if l.hashCode() > r.hashCode() => GreaterThanOrEqual(r, l)
 
-    case Not(GreaterThan(l, r)) if l.hashCode() > r.hashCode() => GreaterThan(r, l)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala]: This is a dead code, because our canonicalization order is bottom up, right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala]: It should be fine to get rid of `assert`, as long as we add the code comments and the needed test cases.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala [hunk 4]
```diff
@@ -32,6 +32,38 @@ class ExpressionSetSuite extends SparkFunSuite {
...
+  // An [AttributeReference] with almost the maximum hashcode, to make testing canonicalize rules
+  // like `case GreaterThan(l, r) if l.hashcode > r.hashcode => GreaterThan(r, l)` easier
+  val maxHash =
+    Canonicalize.ignoreNamesTypes(
+      AttributeReference("maxHash", IntegerType)(exprId =
+        new ExprId(4, NamedExpression.jvmId) {
+          // maxHash's hashcode is calculated based on this exprId's hashcode, so we set this
+          // exprId's hashCode to this specific value to make sure maxHash's hashcode is almost
+          // `Int.MaxValue`
+          override def hashCode: Int = 826929706
```
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala]: Why not `override def hashCode: Int = Int.MaxValue`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala [hunk 5]
```diff
@@ -32,6 +32,38 @@ class ExpressionSetSuite extends SparkFunSuite {
...
+  // An [AttributeReference] with almost the minimum hashcode, to make testing canonicalize rules
+  // like `case GreaterThan(l, r) if l.hashcode > r.hashcode => GreaterThan(r, l)` easier
+  val minHash =
+    Canonicalize.ignoreNamesTypes(
+      AttributeReference("minHash", IntegerType)(exprId =
+        new ExprId(5, NamedExpression.jvmId) {
+          // minHash's hashcode is calculated based on this exprId's hashcode, so we set this
+          // exprId's hashCode to this specific value to make sure minHash's hashcode is almost
+          // `Int.MinValue`
+          override def hashCode: Int = 826929707
```
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala]: Why not `override def hashCode: Int = Int.MinValue`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala]: uh. Just saw the above comment from @dongjoon-hyun . Thanks!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala (hunk 4 above)
Comment (lw-lin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala]: thanks.

the reason is in [Canonicalize.scala#ignoreNamesTypes](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala#L41), we're making copies of `e` (`maxHash` in this case):
```scala
  private def ignoreNamesTypes(e: Expression): Expression = e match {
    case a: AttributeReference =>
      AttributeReference("none", a.dataType.asNullable)(exprId = a.exprId)
    case _ => e
  }
```
so, even if we `override def hashCode: Int = Int.MaxValue` on `maxHash`, it has nothing to do with the copy's hashcode.

then i took a step back -- defined `exprId`'s hashcode to a specific value (as provided in this patch), which would further affect the copied attribute-reference's hashcode.

Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala]: uh, I did not read the comment carefully. Thanks for the explanation. 
You can set it to `-1030353449`. Then, `maxHash.hashCode()` will be equal to `Int.MaxValue`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala (hunk 5 above)
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala]: To make `minHash.hashCode()` equal to `Int.MinValue`, you can set it to `1407330692`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala (hunk 4 above)
Comment (lw-lin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala]: ah, `-1030353449` works great! let me push a commit updating this

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala (hunk 5 above)
Comment (lw-lin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala]: updated, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala]: I think we can remove `assert`, because the test cases already cover the scenario. You can add a comment to explain.

**Comment by gatorsmile** [comment]
LGTM except one comment. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala (hunk 2 above)
Comment (lw-lin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala]: sure!

**Comment by lw-lin** [comment]
Jenkins retest this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala (hunk 4 above)
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSetSuite.scala]: Great!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala [hunk 6]
```diff
@@ -78,13 +78,11 @@ object Canonicalize extends {
...
+    // Note in the following `NOT` cases, `l.hashCode() <= r.hashCode()` holds. The reason is that
+    // canonicalization is conducted bottom-up -- see [[Expression.canonicalized]].
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala]: To the other reviewers, this PR added test cases in `ExpressionSetSuite.scala` to ensure it. Thus, it is safe to clean the codes.

**Comment by gatorsmile** [comment]
Thanks! Merging to master.
